### PR TITLE
reduce DOM size

### PR DESCRIPTION
**Pull request by tvikito**
| Q             | A
| ------------- | ---
|Description, reason for the PR| One of the things which makes our app slower and also get us worse Web Vitals score is DOM size. These changes flattens out the DOM.
|New feature| No 
|[BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

----
:globe_with_meridians: Live Preview:
  - https://tv-fw-1681-reduce-dom-size.odin.shopsys.cloud
  - https://cz.tv-fw-1681-reduce-dom-size.odin.shopsys.cloud

### CODE REVIEW

File: project-base/storefront/components/Basic/Image/Image.tsx [hunk 1]
```diff
@@ -29,20 +40,20 @@ export const Image: FC<ImageProps> = ({ image, alt, type, loading, dataTestId, w
                 data-testid={getDataTestId(dataTestId) + '-empty'}
                 height={height || 160}
                 width={width || 160}
-                className={twMergeCustom('h-auto w-full', classNameTwClass)}
+                className={twMergeCustom('h-auto w-full', imageTwClass, className)}
```
Comment (sebaholesz) [inline project-base/storefront/components/Basic/Image/Image.tsx]: on Line 30, you merge `className` into `imageTwClass` and then you again merge `imageTwClass` with `className`. Not sure if that is correct, or redundant?

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx [hunk 2]
```diff
@@ -1,143 +1,96 @@
-import { AUTOCOMPLETE_CATEGORY_LIMIT, AUTOCOMPLETE_PRODUCT_LIMIT } from './Autocomplete';
+import { AUTOCOMPLETE_CATEGORY_LIMIT, AUTOCOMPLETE_PRODUCT_LIMIT } from './AutocompleteSearchPopup';
```
Comment (sebaholesz) [inline project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: I dont know where in this file I should put it, but your branch throws error when I search for something

image

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx [hunk 3]
```diff
@@ -1,143 +1,96 @@
...
-    );
-
-    const onAutocompleteSearchHandler = useCallback(() => {
-        router.push({ pathname: searchUrl, query: { q: autocompleteSearchQueryValue } });
-    }, [router, autocompleteSearchQueryValue, searchUrl]);
-
-    const onChangeAutocompleteSearchQueryValueHandler: ChangeEventHandler<HTMLInputElement> = useCallback((event) => {
-        setAutocompleteSearchQueryValue(event.currentTarget.value);
-    }, []);
+    const clearSearchQuery = () => setAutocompleteSearchQueryValue('');
```
Comment (sebaholesz) [inline project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: The search query is getting cleared if I focus out :cry:

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx [hunk 4]
```diff
@@ -1,143 +1,96 @@
...
     const [autocompleteSearchQueryValue, setAutocompleteSearchQueryValue] = useState('');
     const debouncedAutocompleteSearchQuery = useDebounce(autocompleteSearchQueryValue, 200);
-    const [hasAutocompleteSearchFocus, setAutocompleteSearchFocus] = useState(false);
+    const { url } = useDomainConfig();
+    const [searchUrl] = getInternationalizedStaticUrls(['/search'], url);
+    const { t } = useTranslation();
+
+    const isWithValidSearchQuery = autocompleteSearchQueryValue.length >= MINIMAL_SEARCH_QUERY_LENGTH;
+
     const [{ data: autocompleteSearchData, fetching: areSearchResultsLoading }] = useAutocompleteSearchQueryApi({
```
Comment (sebaholesz) [inline project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: The search query is not debounced, so it searches after every keystroke.

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx [hunk 5]
```diff
@@ -1,143 +1,96 @@
...
         requestPolicy: 'network-only',
     });
 
-    const autocompleteSearchInRef = useRef<HTMLDivElement>(null);
-    const { url } = useDomainConfig();
-    const [searchUrl] = getInternationalizedStaticUrls(['/search'], url);
-    const { t } = useTranslation();
-    const [isDesktop, setIsDesktop] = useState(false);
-    const { width } = useGetWindowSize();
+    const isSearchResultsPopupVisible = isWithValidSearchQuery && !!autocompleteSearchData;
```
Comment (sebaholesz) [inline project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: This condition makes the popup disappear and appear after every keystroke

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearchPopup.tsx [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+                    {t('View all results')}
+                </Button>
+            </div>
+        </>
+    );
+};
+
+const SearchResultSection: FC = ({ children }) => {
+    return <div>{children}</div>;
+};
```
Comment (sebaholesz) [inline project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearchPopup.tsx]: Not sureif this is necessary?

File: project-base/storefront/components/Blocks/Product/ProductExposedStoresCount.tsx [hunk 7]
```diff
@@ -20,7 +18,7 @@ export const ProductExposedStoresCount: FC<ProductExposedStoresCountProps> = ({
     }
 
     return (
-        <div className={twMergeCustom('text-sm text-black', className)} data-testid={TEST_IDENTIFIER}>
+        <div className="text-sm text-black" data-testid={TEST_IDENTIFIER}>
```
Comment (sebaholesz) [inline project-base/storefront/components/Blocks/Product/ProductExposedStoresCount.tsx]: Why was this removed? I see the classname being passed to this component.

File: project-base/storefront/components/Layout/Header/MenuIconic/MenuIconicItemUserUnauthenticated.tsx [hunk 8]
```diff
@@ -17,22 +17,13 @@ export const MenuIconicItemUserUnauthenticated: FC = ({ dataTestId }) => {
...
 
-            <div className="order-2 ml-1 flex h-9 w-9 cursor-pointer items-center justify-center text-lg outline-none lg:hidden">
-                <div
-                    className="relative flex h-full w-full items-center justify-center text-white transition-colors"
-                    onClick={handleLogin}
-                >
-                    <MenuIconicItemIcon icon={<User />} className="mr-0" />
-                </div>
-            </div>
-
```
Comment (sebaholesz) [inline project-base/storefront/components/Layout/Header/MenuIconic/MenuIconicItemUserUnauthenticated.tsx]: What was this piece of code?

**Review by sebaholesz — CHANGES_REQUESTED** [review]
Nice job :+1: However, it is very hard for me to review all the visual changes, so please make sure to emphasize this for the tester and the PO.

File: project-base/storefront/components/Layout/Header/MenuIconic/MenuIconicItemUserUnauthenticated.tsx (hunk 8 above)
Comment (tvikito) [reply project-base/storefront/components/Layout/Header/MenuIconic/MenuIconicItemUserUnauthenticated.tsx]: Duplication of the same element which is already visible on `lg` viewport and it was only needed to be handled differently in CSS.

File: project-base/storefront/components/Basic/Image/Image.tsx (hunk 1 above)
Comment (tvikito) [reply project-base/storefront/components/Basic/Image/Image.tsx]: Ah, I see. Nice catch. Refactored.

File: project-base/storefront/components/Blocks/Product/ProductExposedStoresCount.tsx (hunk 7 above)
Comment (tvikito) [reply project-base/storefront/components/Blocks/Product/ProductExposedStoresCount.tsx]: Hmm, probably I haven't noticed this one place and it was removed during my refactoring of ProductListItem since there was nothing left to be passed after refactor. Reverted back.

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearchPopup.tsx (hunk 6 above)
Comment (tvikito) [reply project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearchPopup.tsx]: Yup, nothing has left here after refactoring. Refactored.

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx (hunk 4 above)
Comment (tvikito) [reply project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: Fixed now.

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx (hunk 3 above)
Comment (tvikito) [reply project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: Yeah, I haven't though it through. It solved one issue which I had and it didn't see it as bad UX so I left it. But now I see the matrix and I see also this was not good step. It's refactored with red pill now.

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx (hunk 2 above)
Comment (tvikito) [reply project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: After current refactoring I cannot replicate this error.

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx (hunk 5 above)
Comment (sebaholesz) [reply project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: This is still happening (though its not right after the keystroke but when data is being loaded). According to Peta this is a bug. Please, take a look at it.

Comment (sebaholesz) [reply project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx]: this one is solved, nice :)

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx [hunk 9]
```diff
@@ -1,142 +1,117 @@
...
-                                    ? 'max-vl:w-full max-vl:!border-primaryLight'
-                                    : 'border-white',
-                            )}
-                            label={t("Type what you're looking for")}
-                            dataTestId={TEST_IDENTIFIER}
-                            onEnterPressCallback={onAutocompleteSearchHandler}
-                            value={autocompleteSearchQueryValue}
-                            onChange={onChangeAutocompleteSearchQueryValueHandler}
-                            isLoading={areSearchResultsLoading}
+                    {isSearchResultsPopupVisible && (
```
Comment (sebaholesz) [inline project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx:87]: If I click on the autocomplete popup (outside of the links). It disappears.
[Screencast from 30.8.2023 09:16:16.webm](https://github.com/shopsys/shopsys/assets/55181354/7949dcb6-89e8-413a-9046-6850b6c6e80d)

**Review by sebaholesz — CHANGES_REQUESTED** [review]

File: project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx (hunk 9 above)
Comment (tvikito) [reply project-base/storefront/components/Layout/Header/AutocompleteSearch/AutocompleteSearch.tsx:87]: Ah, this is something which I've missed in the implemented logic. It's refactored now.

**Review by sebaholesz — APPROVED** [review]
Nice job
